flax/experimental/nnx/nnx/state.py

```diff
@@ -44,6 +44,58 @@
 FlatState = dict[Path, Variable[Leaf]]
 
 
+class StateVariablesMapping(
+  tp.MutableMapping[str, Variable[tp.Any]], reprlib.Representable
+):
+  __slots__ = ('_mapping',)
+
+  def __init__(self, mapping: dict[Key, tp.Any]):
+    if tp.TYPE_CHECKING:
+      self._mapping = mapping
+    else:
+      object.__setattr__(self, '_mapping', mapping)
+
+  def __getitem__(self, key: str | int) -> Variable[tp.Any]:
+    if isinstance(key, int):
+      key = str(key)
+
+    value = self._mapping[key]
+
+    if not isinstance(value, Variable):
+      raise KeyError(f"Variable '{key}' not found.")
+
+    return value
+
+  def __setitem__(self, name: str, value: Variable[tp.Any]) -> None:
+    self._mapping[name] = value
+
+  def __getattr__(self, name: str) -> Variable[tp.Any]:
+    value = self._mapping[name]
+    if not isinstance(value, Variable):
+      raise AttributeError(f"Variable '{name}' not found.")
+    return value
+
+  def __setattr__(self, name: str, value: Variable[tp.Any]) -> None:
```

Comment:
I would recommend to constrain the API and allow either `["x"]` or `.x` but not both (possibly in a separate PR).

Reply:
We have this dual API in a couple of place in NNX, the main idea is to make intuitive both hard-coded and programmatic access, e.g:

```python
# hard-coded
x = state.linear.kernel
# programmatic
x = state[module_name][param_name]
```